distribution/lib/Standard/Base/0.0.0-dev/src/Data/List.enso

```diff
@@ -402,13 +455,104 @@ type List
+                _ -> case current of
+                    Nil -> Error.throw (Index_Out_Of_Bounds.Error start start-count)
+                    Cons _ b -> @Tail_Call loop b count-1
+            loop self start
+
+    ## Returns the index of the element in the list.
+       Returns Nothing if the element is not found.
+
+       Arguments:
+       - element: The element to search for or a predicate function to test for
+         each element.
+       - start: The index to start searching from. If the index is negative, it
+         is counted from the end of the list.
+
+       > Example
+         Finding a first element of the vector that is larger than 3.
+
+             [1, 2, 3, 4, 5].find (> 3)
+    index_of : (Any | (Any -> Boolean)) -> Integer -> Integer | Nothing
+    index_of self element start=0 = case start.signum of
+        -1 ->
+            start_node = find_start_node self start
+            found = start_node.first.index_of element 0
+            if found.is_nothing then Nothing else found+start_node.second
+        _ ->
+            predicate = if element.is_a Function then element else (==element)
+            loop current count idx = case count of
+                0 -> case current of
+                    Nil -> Nothing
+                    Cons x xs -> if predicate x then idx else @Tail_Call loop xs 0 idx+1
+                _ -> case current of
+                    Nil -> Error.throw (Index_Out_Of_Bounds.Error start start-count)
+                    Cons _ b -> @Tail_Call loop b count-1 idx+1
+            loop self start 0
+
+    ## Returns the last index of the element in the list.
+       Returns Nothing if the element is not found.
+
+       Arguments:
+       - element: The element to search for or a predicate function to test for
+         each element.
+       - start: The index to start searching backwards from. If the index is
+         negative, it is counted from the end of the list.
+
+       > Example
+         Finding a first element of the vector that is larger than 3.
+
+             [1, 2, 3, 4, 5].find (> 3)
+    last_index_of : (Any | (Any -> Boolean)) -> Integer -> Integer | Nothing
+    last_index_of self element start=-1 = case self of
+        Nil -> if start == -1 || start == 0 then Nothing else Error.throw (Index_Out_Of_Bounds.Error start 0)
+        Cons _ _ ->
+            length = self.length
+            used_start = if start < 0 then length+start else start
+            if used_start < 0 || used_start >= length then Error.throw (Index_Out_Of_Bounds.Error start length) else
+                predicate = if element.is_a Function then element else (==element)
+                loop item idx current = if idx>used_start then current else case item of
+                    Nil -> current
+                    Cons x xs ->
+                        new_current = if predicate x then idx else current
+                        @Tail_Call loop xs idx+1 new_current
+                loop self 0 Nothing
```

Comment:
I just realised that the index to start from in this case is to start from backwards, so my suggested code is actually wrong.

I think we can still use the `find_start_node` (for both positive and negative indices, as suggested), just the main loop has to change a bit, probably start from the beginning instead of the `start_node` and stop when it reaches the `offset` index.

Reply:
Don't find this one too complicated - though it is not as efficient as it can be as it computes the length first.

Have left this one unchanged.